dds/DCPS/transport/rtps_udp/RtpsUdpDataLink.cpp

```diff
@@ -1341,85 +1577,233 @@ RtpsUdpDataLink::send_ack_nacks(RtpsReaderMap::value_type& rr, bool finalFlag)
+          response_bundle_addrs.push_back(addr_it->first);
+          response_bundles_sizes.push_back(prev_size + prev_padding);
+          size -= prev_size; padding -= prev_padding; prev_size = 0; prev_padding = 0;
+          prev_dst = GUID_UNKNOWN;
+        } else {
+          prev_size = size; prev_padding = padding;
+        }
+        // Attempt to add the submessage response to the bundle
+        offset = size;
+        Response& res = **resp_it;
+        switch (res.sm_._d()) {
+          case HEARTBEAT: {
+            gen_find_size(res.sm_.heartbeat_sm(), size, padding);
+            res.sm_.heartbeat_sm().smHeader.submessageLength = static_cast<CORBA::UShort>(size - offset) - SMHDR_SZ;
+            break;
+          }
+          case ACKNACK: {
+            gen_find_size(res.sm_.acknack_sm(), size, padding);
+            res.sm_.acknack_sm().smHeader.submessageLength = static_cast<CORBA::UShort>(size - offset) - SMHDR_SZ;
+            break;
+          }
+          case GAP: {
+            gen_find_size(res.sm_.gap_sm(), size, padding);
+            res.sm_.gap_sm().smHeader.submessageLength = static_cast<CORBA::UShort>(size - offset) - SMHDR_SZ;
+            break;
+          }
+          case NACK_FRAG: {
+            gen_find_size(res.sm_.nack_frag_sm(), size, padding);
+            res.sm_.nack_frag_sm().smHeader.submessageLength = static_cast<CORBA::UShort>(size - offset) - SMHDR_SZ;
+            break;
+          }
+          default: {
+            break;
+          }
+        }
+        // If adding the submessage bumped us over the limit, push the size difference into the next bundle, reset prev_dst, and keep going
+        if ((size + padding) > max_bundle_size_) {
+          response_bundles.push_back(ResponseIterVec());
+          response_bundle_addrs.push_back(addr_it->first);
+          response_bundles_sizes.push_back(prev_size + prev_padding);
+          size -= prev_size; padding -= prev_padding; prev_size = 0; prev_padding = 0;
+          prev_dst = GUID_UNKNOWN;
+        } else {
+          prev_size = size; prev_padding = padding;
+        }
+        response_bundles.back().push_back(*resp_it);
+      }
+    }
+  }
+  response_bundles_sizes.push_back(size + padding);
+
+  // Allocate buffers, seralize, and send bundles
+  for (size_t i = 0; i < response_bundles.size(); ++i) {
+    prev_dst = GUID_UNKNOWN;
+    ACE_Message_Block mb_acknack(response_bundles_sizes[i]); //FUTURE: allocators?
+    Serializer ser(&mb_acknack, false, Serializer::ALIGN_CDR);
+    for (ResponseIterVec::const_iterator it = response_bundles[i].begin(); it != response_bundles[i].end(); ++it) {
+      Response& res = **it;
+      RepoId dst = res.dst_guid_;
+      dst.entityId = ENTITYID_UNKNOWN;
+      if (dst != GUID_UNKNOWN && dst != prev_dst) {
+        memcpy(&idst.guidPrefix, dst.guidPrefix, sizeof(idst.guidPrefix));
+        ser << idst;
+        prev_dst = dst;
+      }
+      switch (res.sm_._d()) {
```

Comment:
This is essentially the generated code `operator<<(Serializer&, const Submessage&)`

Reply:
Gotcha. Is there a gen_find_size equivalent? I'm guessing yes... do we want to go ahead and switch this over to try to handle all submessages?

Reply:
I think we might as well use those.  We may not actually support all submessage types in this method but it's fine to use the generated code.

Reply:
I didn't consider find_size for this because the code above also manages the contents of the submessageHeader within the `switch` scope.